### PR TITLE
Fix: Correctly identify Key-User-Actions for web applciations

### PR DESCRIPTION
**Pull request by warber**
#### What this PR does / Why we need it:
KUAs are identifies by name+domain+action type, and not by name only.
Since this information is present in the payload we need a way to compare their values to existing dynatrace objects.

I've introduced a custom function that can be registered on the API level to do this job.
This function can/should be re-used e.g. when dealing with deletion use case.

### CODE REVIEW

File: pkg/client/dtclient/config_client.go [hunk 1]
```diff
@@ -414,11 +414,22 @@ func (d *DynatraceClient) getObjectIdIfAlreadyExists(ctx context.Context, api ap
...
+			var pl map[string]any
+			if err = json.Unmarshal(payload, &pl); err != nil {
+				return "", err
+			}
+			if theAPI.CheckEqualFunc(value.CustomFields, pl) {
 				objectId = value.Id
+				matchingObjectsFound++
+			}
+		} else {
+			if value.Name == objectName || escapeApiValueName(ctx, value) == objectName {
```
Comment (warber) [inline pkg/client/dtclient/config_client.go:427]: this can/should be refactored in it's own CheckEqualFunc.
Maybe a follow up

File: pkg/client/dtclient/config_client.go [hunk 2]
```diff
@@ -605,8 +618,9 @@ func unmarshalJson(ctx context.Context, theApi api.API, resp rest.Response) ([]V
 		}
 		for _, kua := range jsonResp.List {
 			values = append(values, Value{
-				Id:   kua.MeIdentifier,
-				Name: kua.Name,
+				Id:           kua.MeIdentifier,
+				Name:         kua.Name,
+				CustomFields: map[string]any{"name": kua.Name, "domain": kua.Domain, "actionType": kua.ActionType},
```
Comment (warber) [inline pkg/client/dtclient/config_client.go:623]: i thought about populating CustomFields in general and always, however i think it causes too much memory overhead, compared to what we need it for

File: pkg/api/endpoints.go [hunk 3]
```diff
@@ -466,6 +466,11 @@ var configEndpoints = []API{
 		Parent:                       &applicationWebAPI,
 		RequireAllFF:                 []featureflags.FeatureFlag{featureflags.MRumProperties()},
 		TweakResponseFunc:            func(m map[string]any) { delete(m, "meIdentifier") },
+		CheckEqualFunc: func(existing map[string]any, current map[string]any) bool {
+			return existing["name"] == current["name"] &&
+				existing["actionType"] == current["actionType"] &&
+				existing["domain"] == current["domain"]
```
Comment (arthurpitman) [inline pkg/api/endpoints.go:472]: Probably could make `name`, `actionType`, and `domain` constants as we use them a few times

**Comment by warber** [comment]
What I really don't like about my current solution is, that all of that is happening in the scope of the `doXByNAME` function... Maybe I can separate that better

File: pkg/client/dtclient/config_client.go (hunk 2 above)
Comment (UnseenWizzard) [reply pkg/client/dtclient/config_client.go:623]: Agree with this being unnecessary overhead in general - but we could consider if we can use it for any other special case APIs that we already have. e.g. dashboards have some extra fields that are just part of the API value struct now

File: pkg/api/api.go [hunk 4]
```diff
@@ -61,7 +61,10 @@ type API struct {
 	TweakResponseFunc func(map[string]any)
 	// Parent is used for SubPath APIs to store information about the configuration type and ID of the related
 	// configuration once Resolved() is called.
-	Parent *API
+
+	// CheckEqualFunc can be used to compare existing objects with current objects based on their payload
+	CheckEqualFunc func(map[string]any, map[string]any) bool
+	Parent         *API
```
Comment (UnseenWizzard) [inline pkg/api/api.go:67]: you've accidentally removed this field from its comment - should stay on line 64

File: pkg/client/dtclient/config_client.go [hunk 5]
```diff
@@ -414,11 +414,22 @@ func (d *DynatraceClient) getObjectIdIfAlreadyExists(ctx context.Context, api ap
...
 		value := values[i]
-		if value.Name == objectName || escapeApiValueName(ctx, value) == objectName {
-			if matchingObjectsFound == 0 {
+		if theAPI.CheckEqualFunc != nil {
+			var pl map[string]any
+			if err = json.Unmarshal(payload, &pl); err != nil {
+				return "", err
+			}
+			if theAPI.CheckEqualFunc(value.CustomFields, pl) {
 				objectId = value.Id
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/config_client.go:423]: In this case you overwrite the ID if more than one value matches, while so far we'd use the 1st ID (line 418 before, line 428 now) if more than one object match - as we seem to continue with just a warning log, I would not introduce different behavior for APIs with a check equal func and also only take the ID of the first matching object.

File: pkg/client/dtclient/config_client.go [hunk 6]
```diff
@@ -485,7 +411,48 @@ func isUserSessionPropertiesMobile(a api.API) bool {
...
+		values, err := d.fetchExistingValues(ctx, theApi, theApi.CreateURL(d.environmentURLClassic))
+		if err != nil {
+			return "", err
+		}
+		return d.findUniqueByName(ctx, theApi, values, objectName)
+
+	}
+	// The network-zone API doesn't have a POST endpoint, hence, we need to treat it as an update operation
+	// per default
+	if theApi.ID == api.UserActionAndSessionPropertiesMobile {
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/config_client.go]: as these are now all returning directly, can you please structure this method so that they are early exit conditions? 

only if none of these ifs match - and the API is not a `SingleConfiguration` one (make early exit too), do we need to make any API calls to fetch existing values.

Suggestion: 

```go
if theApi.ID == api.UserActionAndSessionPropertiesMobile { return objectName }
if theApi.ID == api.NetworkZone { return objectName }
if theApi.ID == api.CalculatedMetricsLog { return objectName }
if theApi.SingleConfiguration { return "" }

values, err := d.fetchExistingValues(ctx, theApi, theApi.CreateURL(d.environmentURLClassic))
if err != nil {
	return "", err
}

if theApi.CheckEqualFunc != nil { return d.FindUnique } 

return d.FindUniqueByName
```

File: pkg/client/dtclient/config_client.go [hunk 7]
```diff
@@ -42,6 +42,24 @@ func (d *DynatraceClient) upsertDynatraceObject(ctx context.Context, theApi api.
 			return DynatraceEntity{}, err
 		}
 
+		// The network-zone API doesn't have a POST endpoint, hence, we need to treat it as an update operation
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/config_client.go:45]: why did you move these back out after the fetching on IDs?

File: pkg/client/dtclient/config_client.go [hunk 8]
```diff
@@ -627,24 +627,35 @@ func unmarshalJson(ctx context.Context, theApi api.API, resp rest.Response) ([]V
...
-				Domain       string `json:"domain"`
-				ActionType   string `json:"actionType"`
-			} `json:"keyUserActionList"`
-		}
+		var jsonResp map[string]any
 		err := json.Unmarshal(resp.Body, &jsonResp)
+
+		kuaL, ok := jsonResp[api.KeyUserActionsListField].([]any)
+		if !ok {
+			return nil, fmt.Errorf("unable to find field %s in key user alction list response", api.KeyUserActionsListField)
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/config_client.go:635]: `alction` typo in each `action`

File: pkg/client/dtclient/config_client.go (hunk 7 above)
Comment (warber) [reply pkg/client/dtclient/config_client.go:45]: because they are in the end deciding about whether an update or a create will be executed...
having that early return in another function hides that side effect entirely

File: pkg/client/dtclient/config_client.go [hunk 9]
```diff
@@ -573,6 +540,62 @@ func (d *DynatraceClient) getExistingValuesFromEndpoint(ctx context.Context, the
...
+		}
+		if checkEqualFunc(value.CustomFields, pl) {
+			if matchingObjectsFound == 0 {
+				objectId = value.Id
+			}
+			matchingObjectsFound++
+		}
+	}
+
+	return objectId, nil
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/config_client.go:596]: you're now missing the warning and debug logs that the findUniqueByName has on lines 556-562. I think it's worth having that information in the logs for custom equals functions too

File: pkg/client/dtclient/config_client_test.go [hunk 10]
```diff
@@ -408,14 +408,12 @@ func Test_getObjectIdIfAlreadyExists(t *testing.T) {
...
 
 			dtclient, _ := NewDynatraceClientForTesting(server.URL, server.Client(), nil)
-			got, err := dtclient.getObjectIdIfAlreadyExists(context.TODO(), testApi, server.URL, tt.givenObjectName, nil)
+
+			_, _, err := dtclient.configExistsByName(context.TODO(), testApi, tt.givenObjectName)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("getObjectIdIfAlreadyExists() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if got != tt.wantFoundId {
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/config_client_test.go:416]: why did you remove this check? If it is really not needed please remove the `wantFoundId` field from the testcases =

Comment (warber) [reply pkg/client/dtclient/config_client_test.go]: accidentally

**Review by UnseenWizzard — APPROVED** [review]
thx for incorporating all comments

**Review by arthurpitman — APPROVED** [review]
